Better handling of thousand separators

danielhollas opened the pull request:
This PR has two goals:

1. Safer handling of thousand separator regex
2. More rules for normalizeMathTranslations so that we are less strict about the thousand separators.

Ad. 1. 
_Strategy_: 
As a very first regex, replace thousand separators with placeholders, to reduce risk of mingling with decimal commas. This will also allow us to better handle languages which have decimal commas and a dot for thousand separator (i.e. reversed EN notation).
_Challenge_: 
For the first regex, we will need to combine locale lists for all different notations. This means creating more consts like `THOUSAND_SEPARATOR_AS_THIN_SPACE`. At this point, it might be better to encapsulate them in an object.

danielhollas commented:
Okay, I think the main code is ready and tests were added. Would appreciate feedback before I go clean up the code.

mpolyak commented on lib/translation-assistant.js:
```diff
@@ -355,9 +355,17 @@ function createTemplate(englishStr, translatedStr, lang) {
+// TODO: Need to extract this out of tests and use MATH_LOCALES instead
 var THOUSAND_SEPARATOR_AS_THIN_SPACE_LOCALES = ['cs', 'fr', 'de', 'pt-pt', 'nb', 'bg', 'pl', 'ro', 'nl', 'az', 'sv', 'it', 'hu', 'uk'];
 
+var MATH_LOCALES = {
```
Is the intent behind `MATH_LOCALES` is to ease exporting these rules to the tests, and to communicate that they're related concepts?

If so, I suggest renaming to something like `MATH_LOCALES_RULES`, and upper case the the properties to indicate that they're constant:

```
const MATH_LOCALES_RULES = {
    NO_THOUSAND_SEP: ['ko', 'ps'],
    ...
};
```

danielhollas replied:
Okay, renaming definitely make sense, thanks. I wasn't sure about using upper case in keys, will do.

The most important reason is that these are used in another function: `normalizeTranslatedMath` so these needs to be separated. And at this point it prob. make sense to do it to all locale lists, even though not all of them are needed in that second function.
And, yes, this will make testing easier.

mpolyak commented on src/translation-assistant.js:
```diff
@@ -379,7 +388,25 @@ const THOUSAND_SEPARATOR_AS_THIN_SPACE_LOCALES = ['cs', 'fr', 'de',
  */
 function translateMath(math, lang) {
 
+    const placeholder = 'THSEP';
```
Any reason not to extract these to file level as constants?

danielhollas replied:
Hmm, this placeholder is used only in this function (at least at this point) so I don't think it needs to be accessible outside. Am I missing something?

danielhollas replied:
Maybe I am misunderstanding what do you mean by "these". I'm definitely gonna extra locale lists from `mathTranslations`

danielhollas commented:
@mpolyak this is now good to go from my side. Once you merge #14, I will need to carefully rebase this.

mpolyak commented on src/translation-assistant.js:
```diff
@@ -358,38 +358,69 @@ function createTemplate(englishStr, translatedStr, lang) {
- * TODO(danielhollas): For now only the obvious cases were implemented.
  * TODO(danielhollas): Need to update this when new langs join translations
+ */
+const MATH_RULES_LOCALES = {
```
Let's drop the quotes from the around the properties as they're unneeded:

```
const MATH_RULES_LOCALES = {
    THOUSAND_SEP_AS_THIN_SPACE: [...],
    THOUSAND_SEP_AS_DOT: [...],
    ...
};
```

mpolyak commented on src/translation-assistant.js:
```diff
@@ -358,38 +358,69 @@ function createTemplate(englishStr, translatedStr, lang) {
+    const USThousandSeparatorRegex =
+        new RegExp(`([0-9])${PLACEHOLDER}([0-9])(?=[0-9]{2})`, 'g');
+    const thousandSeparatorLocales =
+            MATH_RULES_LOCALES['THOUSAND_SEP_AS_THIN_SPACE']
```
Let's change how these locales rules are accessed, .eg:

```
const thousandSeparatorLocales = MATH_RULES_LOCALES.THOUSAND_SEP_AS_THIN_SPACE.concat(...`
```

mpolyak commented on tests/assistant_spec.js:
```diff
@@ -489,12 +489,32 @@ describe('translateMath', function() {
         const translatedStr = '1\\,000\\,000 + 9\\,000';
 
-        THOUSAND_SEPARATOR_AS_THIN_SPACE_LOCALES.forEach(function(locale) {
+        MATH_RULES_LOCALES['THOUSAND_SEP_AS_THIN_SPACE']
```
Here as well: `MATH_RULES_LOCALES.THOUSAND_SEP_AS_THIN_SPACE`

mpolyak commented on src/translation-assistant.js:
```diff
@@ -441,18 +464,26 @@ function translateMath(math, lang) {
+         // We cannot allow a literal space here, cause Tex would ignore it
+         {langs: MATH_RULES_LOCALES['THOUSAND_SEP_AS_THIN_SPACE'],
+            regex: /([0-9])\{?~\}?([0-9])(?=[0-9]{2})/g, replace: '$1\\,$2'},
+         //TODO: Strip extra braces around a dot as a thousand separator
```
Is this going to revisited in #14?

I'm concerned leaving commented code with a question like that. As context will be lost over time. If this is an open question I would suggest deleting this block and creating a JIRA issue on the IC board with all relevant information.

danielhollas replied:
Nope, #14 should precede this one. 

Yeah, this is kind of an open question, I will open it up on i18n forum with the relevant teams. I can delete the code, although I'd like to leave a NOTE at least about it. Would that be okay. In any case, don't plan on leaving this undecided for long.

mpolyak replied:
Yep a NOTE makes sense, thank you.

mpolyak commented on src/translation-assistant.js:
```diff
@@ -358,38 +358,69 @@ function createTemplate(englishStr, translatedStr, lang) {
  */
 function translateMath(math, lang) {
 
+    const PLACEHOLDER = 'THSEP';
```
Could you please add a comment for these function constants, mentioning that these are local to the function, regexs that are used to manipulate expressions?

Let's also lowecase `PLACEHOLDER` please, as it's not a file const (it looks like it should be a file level const, which confuses me every time I see it :)

danielhollas commented:
Okay, I've rebased this, but decided to solve the issue with extra brackets before merging. I contacted the relevant teams and gave them a week to reply.

mpolyak commented:
Ok @danielhollas, I will wait for your go ahead once you hear back.

danielhollas commented:
@mpolyak Okay, I've heard back from enough people and it seems that stripping extra braces is fine. I've modified/uncommented the code and added one more test or `normalizeTranslatedMath`. This is now good to go from my side. :crossed_fingers:

mpolyak commented:
@danielhollas, I'll see about deploying both changes this week.
